core/server/policy_violations.go

```diff
@@ -0,0 +1,263 @@
+				}
+			}
+		}
+	}
+	labelSelector, err := k8sLabels.ValidatedSelectorFromSet(map[string]string{
+		"pac.weave.works/type": "Admission"})
+	if err != nil {
+		return nil, fmt.Errorf("error building selector for events query: %v", err)
+	}
+
+	fieldSelectorSet := map[string]string{
+		"type": "Warning",
+	}
+
+	if m.Application != "" {
+		fieldSelectorSet["involvedObject.name"] = m.Application
+		fieldSelectorSet["involvedObject.kind"] = m.Kind
+	}
+
+	if m.Namespace != "" {
+		fieldSelectorSet["involvedObject.namespace"] = m.Namespace
+	}
+
+	fieldSelector := k8sFields.SelectorFromSet(fieldSelectorSet)
+
+	opts := []sigsClient.ListOption{}
+	if m.Pagination != nil {
+		opts = append(opts, sigsClient.Limit(m.Pagination.PageSize))
+		opts = append(opts, sigsClient.Continue(m.Pagination.PageToken))
+	}
+	opts = append(opts, &sigsClient.ListOptions{
+		LabelSelector: labelSelector,
+		FieldSelector: fieldSelector,
+	})
+	opts = append(opts, sigsClient.InNamespace(v1.NamespaceAll))
+
+	validationsList, err := cs.listEvents(ctx, clustersClient, m.ClusterName, false, opts)
+	if err != nil {
+		return nil, fmt.Errorf("error getting events: %v", err)
+	}
+	respErrors = append(respErrors, validationsList.Errors...)
+	policyviolationlist := pb.ListPolicyValidationsResponse{
+		Total:         int32(len(validationsList.Validations)),
+		Violations:    validationsList.Validations,
+		Errors:        respErrors,
+		NextPageToken: validationsList.Token,
+	}
+	return &policyviolationlist, nil
+}
+
+func (cs *coreServer) GetPolicyValidation(ctx context.Context, m *pb.GetPolicyValidationRequest) (*pb.GetPolicyValidationResponse, error) {
+	clusterClient, err := cs.clustersManager.GetImpersonatedClient(ctx, auth.Principal(ctx))
+	if err != nil {
+		return nil, fmt.Errorf("error getting impersonating client: %w", err)
+	}
+
+	selector, err := k8sLabels.ValidatedSelectorFromSet(map[string]string{
+		"pac.weave.works/type": "Admission",
+		"pac.weave.works/id":   m.ViolationId})
+
+	if err != nil {
+		return nil, fmt.Errorf("error building selector for events query: %v", err)
+	}
+	opts := []sigsClient.ListOption{}
+
+	fields := k8sFields.OneTermEqualSelector("type", "Warning")
+	opts = append(opts, &sigsClient.ListOptions{
+		LabelSelector: selector,
+		FieldSelector: fields,
+	})
+	opts = append(opts, sigsClient.InNamespace(v1.NamespaceAll))
+
+	validationsList, err := cs.listEvents(ctx, clusterClient, m.ClusterName, true, opts)
+	if err != nil {
+		return nil, fmt.Errorf("error getting events: %v", err)
+	}
+	if len(validationsList.Errors) > 0 {
+		return nil, fmt.Errorf("error getting events: %s", validationsList.Errors[0].Message)
+	}
+	if len(validationsList.Validations) == 0 {
+		return nil, fmt.Errorf("no policy violation found with id %s and cluster: %s", m.ViolationId, m.ClusterName)
+	}
+	return &pb.GetPolicyValidationResponse{
+		Violation: validationsList.Validations[0],
```

Comment:
This is a little confusing. We only care about the first validation?

Reply:
The method purpose is to get the details of one violation, yet the method we query with `listEvents` (used in list&get)
returns a list, so we only need that one object

There also safety checks [here]( https://github.com/weaveworks/weave-gitops//blob/ff6ff54f6e376e916e8cb52390f93d22b6ee37d5/core/server/policy_violations.go#L113) if something went wrong and the query returned none or more than object we return some errors